README.md
fuse-vdv
=======

This tool provides a FUSE based filesystems that enables you to virtually 
concatenate a sequence of files. Given the purpose of the tool, it works 
best with video files in "raw DV" format.

The tool is tuned for the needs of the FeM/VOC CRS project (URL to come 
once it's publicly released) which contains some more documentation.



debian/changelog
fuse-vdv (2016.03.25.1) stable; urgency=low

  * version bump: relative path in shotcut file, inherently supporting
    Shotcut on all platforms

 -- Matthias Hunstock <atze@fem.tu-ilmenau.de>  Fri, 25 Mar 2016 20:40:23 +0100

fuse-vdv (2015.03.12.2) stable; urgency=low

  * Improvement: better error reporting in shotcut support

 -- Matthias Hunstock <atze@fem.tu-ilmenau.de>  Thu, 12 Mar 2015 10:39:57 +0100

fuse-vdv (2015.03.12.1) stable; urgency=high

  * Bugfix: 32bit calculations produce error if source pieces are bigger than 4 GB

 -- Matthias Hunstock <atze@fem.tu-ilmenau.de>  Thu, 12 Mar 2015 09:30:23 +0100

fuse-vdv (2014.12.14.1) stable; urgency=low

  * Bugfixing in KDEnlive project files

 -- Matthias Hunstock <atze@fem.tu-ilmenau.de>  Sun, 14 Dec 2014 23:32:45 +0100

fuse-vdv (2014.09.07.1) stable; urgency=low

  * Refactor and bugfix support for shotcut project files (bjoern)
  * Bugfixing in KDEnlive project files

 -- Matthias Hunstock <atze@fem.tu-ilmenau.de>  Sun, 07 Sep 2014 22:00:03 +0200

fuse-vdv (2014.08.20.1) stable; urgency=low

  * Add support for shotcut project files (bjoern)
  * Remove parsing of timestamps in filenames

 -- Matthias Hunstock <atze@fem.tu-ilmenau.de>  Wed, 20 Aug 2014 19:48:02 +0200

fuse-vdv (2014.07.28.1) stable; urgency=low

  * Fix build issues on Ubuntu

 -- Matthias Hunstock <atze@fem.tu-ilmenau.de>  Mon, 28 Jul 2014 14:45:19 +0200

fuse-vdv (2014-07-13-1) stable; urgency=low

  * Enhancement: log errors to STDERR before backgrounding

 -- Matthias Hunstock <atze@fem.tu-ilmenau.de>  Sun, 13 Jul 2014 11:58:23 +0200

fuse-vdv (2014-05-27-1) stable; urgency=low

  * Enhancement: improved support for single source files by making prefix really optional

 -- Matthias Hunstock <atze@fem.tu-ilmenau.de>  Tue, 27 May 2014 15:37:07 +0200

fuse-vdv (2014-04-28-1) stable; urgency=low

  * Improved error messages
  * Improved checking of error conditions or bad/empty parameters
  * always log to stderr

 -- Peter Koerner  <peter@mazdermind.de>  Sun, 20 Apr 2014 20:06:17 +0200

fuse-vdv (2014-04-19-1) stable; urgency=low

  * Bugfix: Reopening KDEnlive project files failed
  * Enhancement: Empty space on KDEnlive track is preserved in virtual project file

 -- Matthias Hunstock <atze@fem.tu-ilmenau.de>  Sat, 19 Apr 2014 09:43:17 +0100

fuse-vdv (2013-12-13-1) stable; urgency=low

  * Multithread support

 -- Matthias Hunstock <atze@fem.tu-ilmenau.de>  Fri, 13 Dec 2013 22:23:46 +0100



fuse-vdv-tools.h
// some sane default... the longest stuff are filenames
static const size_t max_str_len = 8192;

///////////////// string stuff  ///////////////

extern char * dupe_str (const char * source);
extern char * dupe_str_n (const char * source, size_t count);
extern char * merge_str (const char ** sources, size_t count);
extern char * merge_strs (int count, ...);
extern char * append_and_free (char * s1, char * s2);
extern char * update_int_string (char * str, int value, size_t* length);
extern size_t write_to_string (const char * buffer, size_t size, off_t offset, char ** target);
extern size_t safe_strlen (const char * str);
extern size_t string_read (const char * str, char * buf, size_t size, const off_t offset);
extern size_t string_read_with_length (const char * str, char * buf, size_t size, const off_t offset, size_t length);
extern char * get_datestring_from_filename(const char * filename);
extern char * equalize_date_string (const char * s);
extern int    compare_date_strings (const char * s1, const char * s2);
extern int    datestring_to_timestamp (const char * s);

extern size_t truncate_buffer (char ** target, size_t bufferlength, size_t newbufferlength);

// TODO: eliminate:
extern size_t write_to_buffer (const char *buffer, size_t size, off_t offset, char **target, size_t * bufferlength);

///////////////// file stuff  ///////////////

extern int file_exists (const char * filename);

///////////////// memory stuff  ///////////////

#include <err.h>
#define CHECK_OOM(pointer) if (pointer == NULL) err (123, "Out of memory! (function: %s) Exiting!\n", __FUNCTION__);



fuse-vdv-shotcut.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <inttypes.h>
#include <fcntl.h>
#include <sys/types.h>
#include <assert.h>
#include <mxml.h>
#include <pthread.h>
#include "fuse-vdv.h"
#include "fuse-vdv-tools.h"
#include "fuse-vdv-debug.h"
#include "fuse-vdv-filebuffer.h"
#include "fuse-vdv-shotcut.h"

const char *shotcut_path = "/project_shotcut.mlt";
static const char *sc_template;

static filebuffer_t* sc_project_file_cache = NULL;
static int sc_project_file_cache_frames = -1;
static int sc_project_file_refcount = 0;
static pthread_mutex_t sc_cachemutex = PTHREAD_MUTEX_INITIALIZER;

static filebuffer_t *sc_writebuffer = NULL;

filebuffer_t* get_shotcut_project_file_cache (const char *filename, int num_frames, int blanklen) {
	pthread_mutex_lock (&sc_cachemutex);

	if ((sc_project_file_cache != NULL) && (sc_project_file_cache_frames == num_frames)) {
		debug_printf ("get_shotcut_project_file: filename: '%s' frames: %d --> cache hit: (%p)\n", filename, num_frames, sc_project_file_cache);
		pthread_mutex_unlock (&sc_cachemutex);
		return sc_project_file_cache;
	}

	int _outframe = (outframe < 0) ? lastframe : outframe;
	const size_t size = strlen(sc_template) * 2;
	if (sc_project_file_cache == NULL) sc_project_file_cache = filebuffer__new();
	char* temp = (char *) malloc (size);
	CHECK_OOM(temp);
	int len = snprintf (temp, size - 1, sc_template, inframe, num_frames, num_frames - 1, (off_t) (num_frames * frame_size), filename, _outframe, blanklen);
	if (len >= size) err(124, "%s: size fail when generating project file\n", __FUNCTION__);
	debug_printf ("get_shotcut_project_file: result has a size of: %d\n", len);
	filebuffer__write(sc_project_file_cache, temp, len, 0);
	filebuffer__truncate(sc_project_file_cache, len);
	sc_project_file_cache_frames = num_frames;
	free (temp);

	pthread_mutex_unlock (&sc_cachemutex);
	return sc_project_file_cache;
}

size_t get_shotcut_project_file_size (const char *filename, int num_frames, int blanklen) {
	filebuffer_t* fb = get_shotcut_project_file_cache (filename, num_frames, blanklen);
	return filebuffer__contentsize(fb);
}

void init_shotcut_project_file () {
	pthread_mutex_lock (&sc_cachemutex);
	if (sc_project_file_cache != NULL) {
		debug_printf ("init_shotcut_project_file: freeing cache pointer %p\n", sc_project_file_cache);
		filebuffer__destroy (sc_project_file_cache);
		sc_project_file_cache = NULL;
	}
	pthread_mutex_unlock (&sc_cachemutex);
}

size_t shotcut_read (const char *path, char *buf, size_t size, off_t offset, const char *movie_path, int frames, int blanklen) {
	debug_printf ("reading from shotcut project file at %" PRId64 " with a length of %d \n", offset, size);
	filebuffer_t* fb = get_shotcut_project_file_cache (movie_path + 1, frames, blanklen);
	if (fb == NULL) return -EIO;
	return filebuffer__read (fb, offset, buf, size);
}

void open_shotcut_project_file (const char *movie_path, int frames, int blanklen, int truncate) {
	debug_printf ("opening shotcut project file, truncate:%d\n", truncate);
	sc_project_file_refcount++;
	if (sc_project_file_refcount > 1) return;
	if (sc_writebuffer == NULL) {
		debug_printf ("creating new writebuffer from project file\n");
		if (truncate) {
			sc_writebuffer = filebuffer__new();
		} else {
			sc_writebuffer = filebuffer__copy(get_shotcut_project_file_cache (movie_path, frames, blanklen));
		}
	} else if (truncate) {
		filebuffer__truncate(sc_writebuffer, 0);
	}
}

int truncate_shotcut_project_file(size_t size) {
	if (sc_writebuffer != NULL)  {
		size_t l = filebuffer__truncate(sc_writebuffer, size);
		if (l < 0 || l < size) {
			return -EIO;
		}
	} else {
		sc_writebuffer = filebuffer__new();
	}
	return 0;
}

size_t write_shotcut_project_file (const char *buffer, size_t size, off_t offset) {
	debug_printf ("writing to shotcut project file at %" PRId64 " with a length of %d \n", offset, size);
	if (sc_writebuffer == NULL) {
		debug_printf ("writing to shotcut project FAILED: not opened before!\n");
		return -EACCES;
	}
	return filebuffer__write(sc_writebuffer, buffer, size, offset);
}

void close_shotcut_project_file () {
	debug_printf ("closing shotcut project file.\n");
	sc_project_file_refcount--;
	if (sc_project_file_refcount > 0) return;
	// TODO save projectfile and hand it out as a different file
	if (sc_writebuffer != NULL) {
		filebuffer__destroy(sc_writebuffer);
		sc_writebuffer = NULL;
	}
}

int find_cutmarks_in_shotcut_project_file (int *inframe, int *outframe, int *blanklen) {
	if (sc_writebuffer == NULL) {
		debug_printf ("find_cutmarks: file has not been written to.\n");
		return 100;
	}
/*
  in XPATH, I would look for 
    producer[@id='producer0']/@in
  and
    producer[@id='producer0']/@out
*/
	mxml_node_t *xmldoc;
	char* temp = filebuffer__read_all_to_cstring(sc_writebuffer);
	xmldoc = mxmlLoadString (NULL, temp, MXML_TEXT_CALLBACK);
	free(temp);
	if (NULL == xmldoc) {
		debug_printf ("find_cutmarks: no valid XML!\n");
		return 1;
	}
	mxml_node_t *node;
	node = mxmlFindElement (xmldoc, xmldoc, "producer", "id", "producer0", MXML_DESCEND);
	if (NULL == node) {
		debug_printf ("find_cutmarks: node with id 'producer0' not found!\n");
		mxmlRelease (xmldoc);
		return 2;
	}

	int blank = 0;
	const char *strin = mxmlElementGetAttr (node, "in");
	const char *strout = mxmlElementGetAttr (node, "out");

	if (NULL == strin) {
		debug_printf ("find_cutmarks: no valid inpoint found!\n");
		mxmlRelease (xmldoc);
		return 4;
	}
	if (NULL == strout) {
		debug_printf ("find_cutmarks: no valid outpoint found!\n");
		mxmlRelease (xmldoc);
		return 5;
	}

	debug_printf ("find_cutmarks: found attributes in='%s' out='%s'\n", strin, strout);

	int in, out = 0;
	if (index(strin, ':') != NULL) {
		static int h,m,s,ms = 0;
		if ((4 != sscanf(strin,"%d:%d:%d,%d",&h,&m,&s,&ms)) && (4 != sscanf(strin,"%d:%d:%d.%d",&h,&m,&s,&ms))) {
			error_printf ("find_cutmarks: no valid inpoint found!\n");
			mxmlRelease (xmldoc);
			return 4;
		}
		in = (h*60*60*frames_per_second)+(m*60*frames_per_second)+(s*frames_per_second)+(ms/(1000/frames_per_second));
	} else {
		in = atoi (strin);
	}

	if (index(strout, ':') != NULL) {
		static int h,m,s,ms = 0;
		if ((4 != sscanf(strout,"%d:%d:%d,%d",&h,&m,&s,&ms)) && (4 != sscanf(strout,"%d:%d:%d.%d",&h,&m,&s,&ms))) {
			error_printf ("find_cutmarks: no valid outpoint found!\n");
			mxmlRelease (xmldoc);
			return 5;
		}
		out = (h*60*60*frames_per_second)+(m*60*frames_per_second)+(s*frames_per_second)+(ms/(1000/frames_per_second));

	} else {
		out = atoi (strout);
	}
	
	mxmlRelease (xmldoc);
	if (0 > in) {
		error_printf ("find_cutmarks: inpoint invalid!\n");
		return 6;
	}
	if (0 >= out) {
		error_printf ("find_cutmarks: outpoint invalid!\n");
		return 7;
	}
	debug_printf ("find_cutmarks: blank is '%d'\n", blank);
	debug_printf ("find_cutmarks: in is '%d'\n", in);
	debug_printf ("find_cutmarks: out is '%d'\n", out);
	*blanklen = blank;
	*inframe = in;
	*outframe = out;
	init_shotcut_project_file();
	return 0;
}


//   %1$d => inframe,  %2$d => frames, %3$d => frames - 1 
//   %4$(PRI64d) => filesize, %5$s => filename without path
//   %6$d => outframe

static const char *sc_template =
"<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
"<mlt LC_NUMERIC=\"C\" version=\"0.9.9\" title=\"FUSE-TS\" parent=\"producer0\" in=\"%1$d\" out=\"%6$d\">\n"
"  <profile description=\"automatic\" />\n"
"  <producer id=\"producer0\" title=\"Source Clip\" in=\"%1$d\" out=\"%6$d\">\n"
"    <property name=\"mlt_type\">mlt_producer</property>\n"
"    <property name=\"length\">%2$d</property>\n"
"    <property name=\"eof\">pause</property>\n"
"    <property name=\"resource\">%5$s</property>\n"
"    <property name=\"seekable\">1</property>\n"
"    <property name=\"aspect_ratio\">1,06667</property>\n"
"    <property name=\"audio_index\">1</property>\n"
"    <property name=\"video_index\">0</property>\n"
"    <property name=\"mute_on_pause\">0</property>\n"
"    <property name=\"mlt_service\">avformat-novalidate</property>\n"
"    <property name=\"ignore_points\">0</property>\n"
"    <property name=\"global_feed\">1</property>\n"
"  </producer>\n"
"  <!--\n"
"  %1$d => inframe,  %2$d => frames, %3$d => frames - 1\n"
"  %4$" PRId64 " => filesize, %5$s => filename without path\n"
"  %6$d => outframe %7$d => blanklen\n"
"  -->\n"
"</mlt>\n";



fuse-vdv-tools.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <errno.h>
#include <fcntl.h>
#include <math.h>
#include <dirent.h>
#include <assert.h>
#include <inttypes.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <time.h>
#include "fuse-vdv.h"
#include "fuse-vdv-tools.h"
#include "fuse-vdv-debug.h"

// function from STL

char *strptime(const char *buf, const char *format, struct tm *tm);

char *dupe_str (const char *source) {
	size_t L = safe_strlen (source);
	if (L > max_str_len) {
		fprintf (logging, "Exception: string that is too long - aborting\n");
		exit (1000);
	}
	char *ret = malloc (L + 1);
	memcpy (ret, source, L);
	ret[L] = 0;
	return ret;
}

char *dupe_str_n (const char *source, size_t count) {
	size_t L = safe_strlen (source);
	if (L > count) {
		L = count;
	}
	char *ret = malloc (L + 1);
	memcpy (ret, source, L);
	ret[L] = 0;
	return ret;
}

char *merge_str (const char **sources, size_t count) {
	int i = 0;
	size_t LL[count];
	size_t L = 0;
	for (; i < count; i++) {
		size_t l = safe_strlen (sources[i]);
		if (l > max_str_len) {
			fprintf (logging, "Exception: string that is too long - aborting\n");
			exit (1002);
		}
		LL[i] = l;
		L += l;
	}
	char *ret = malloc (L + 1);
	size_t o = 0;
	for (i = 0; i < count; i++) {
		if (sources[i] != NULL) {
			memcpy (ret + o, sources[i], LL[i]);
			o += LL[i];
		}
	}
	ret[L] = 0;
	return ret;
}

char *merge_strs (int count, ...) {
	if (count == 0) {
		return NULL;
	}
	va_list args;
	va_start (args, count);
	//char ** t = (char**) malloc(count * sizeof(char*));
	const char *t[count];
	int i = 0;
	for (; i < count; i++) {
		t[i] = va_arg (args, char *);
	}
	va_end (args);
	char *ret = merge_str (t, count);
//  free(t);
	return ret;
}

char *append_and_free (char *s1, char *s2) {
	size_t L1 = safe_strlen (s1);
	size_t L2 = safe_strlen (s2);
	if (L1 + L2 > max_str_len) {
		fprintf (logging, "Exception: string that is too long - aborting\n");
		exit (1002);
	}
	char *ret = malloc (L1 + L2 + 1);
	memcpy (ret, s1, L1);
	memcpy (ret + L1, s2, L2);
	ret[L1 + L2] = 0;
	free (s1);
	free (s2);
	return ret;
}

char *update_int_string (char *str, int value, size_t * length) {
	char *ret = (char *) malloc (16);
	int l = snprintf (ret, 16, "%d\n", value);
	if (l > 14) {
		debug_printf ("update_int_string: FAIL: int too long (more than 14 characters) ???\n");
		ret[0] = 0;
		l = 0;
	}
	if (str != NULL) {
		free (str);
	}
	if (length != NULL) {
		*length = l;
	}
	return ret;
}

size_t write_to_string (const char *buffer, size_t size, off_t offset, char **target) {
	if ((target == NULL) || (*target == NULL)) {
		debug_printf ("writing string failed: NULL target!\n");
		return -EACCES;
	}
	size_t writebuffersize = strlen (*target);
	return write_to_buffer (buffer, size, offset, target, &writebuffersize);
}

size_t safe_strlen (const char *str) {
	if (str == NULL) {
		return 0;
	}
	return strlen (str);
}

size_t truncate_buffer (char ** target, size_t bufferlength, size_t newbufferlength) {
	// let shrinking happen without reallocation, growing will trigger new malloc, even if it would fit
	if (newbufferlength <= bufferlength) {
		return newbufferlength;
	}
	char *old = *target;
	*target = (char *) calloc (1, newbufferlength);
	if (!*target) {
		// malloc failed
		*target = old;
		return -EACCES;
	}
	memcpy (*target, old, bufferlength);
	free (old);
	return newbufferlength;
}

size_t write_to_buffer (const char *buffer, size_t size, off_t offset, char **target, size_t * bufferlength) {
	DEPRECATED();
	debug_printf ("write_to_buffer:  writing %d bytes to %p with offset %" PRId64 " and bufferlength %d\n", size, *target, offset, *bufferlength);
	if ((target == NULL) || (bufferlength == NULL)) {
		debug_printf ("writing buffer failed: NULL pointer!\n");
		return -EACCES;
	}
	size_t writebuffersize = (*bufferlength);
	if (*target == NULL) {
		debug_printf ("write_to_buffer: creating new buffer for NULL target\n");
		writebuffersize = offset + size + 1;
		*target = (char *) malloc (writebuffersize);
		memset (*target, 0, writebuffersize);
		writebuffersize--;
	} else if ((offset + size) > writebuffersize) {
		writebuffersize = offset + size + 1;
		size_t l = truncate_buffer(target, *bufferlength, writebuffersize);
		if (!l || l < writebuffersize) {
			error_printf ("truncating buffer failed!\n");
			return -EACCES;
		}
		writebuffersize--;
	}
	assert (writebuffersize >= offset + size);
	char *dest = *target;
	memcpy (dest + offset, buffer, size);
	*bufferlength = writebuffersize;
	return size;
}

size_t string_read (const char *str, char *buf, size_t size, const off_t offset) {
	size_t length = safe_strlen (str);
	if ((length == 0) || (offset >= length) || (offset < 0)) {
		return 0;
	}
	return string_read_with_length (str, buf, size, offset, length);
}

size_t string_read_with_length (const char *str, char *buf, size_t size, const off_t offset, size_t length) {
	if ((str == NULL) || (length == 0) || (offset >= length) || (offset < 0)) {
		return 0;
	}
	assert (offset >= 0);
	if ((offset + size) > length) {
		size = length - offset;
	}
	memcpy (buf, str + offset, size);
	return size;
}

char *get_datestring_from_filename (const char *filename) {
	char *ret = dupe_str (filename);
	if (filename == NULL) {
		return ret;
	}
	const char *dot = strrchr (filename, '.');
	size_t s;
	if (dot == NULL) {
		s = safe_strlen (filename);
	} else {
		if (safe_strlen (filename) - (dot - filename) <= 5) {
			s = dot - filename;
		} else {
			s = safe_strlen (filename);
		}
	}
	if (s <= 18 || s > 1024) {
		return ret;
	}
	free (ret);
	return dupe_str_n ((const char *) (filename + s - 19), 19);
}

char *equalize_date_string (const char *s) {
	char *ret = dupe_str (s);
	if (safe_strlen (s) > 16) {
		ret[4] = 0x20;
		ret[7] = 0x20;
		ret[10] = 0x20;
		ret[13] = 0x20;
		ret[16] = 0x20;
	}
	return ret;
}

int compare_date_strings (const char *s1, const char *s2) {
	char *t1 = equalize_date_string (s1);
	char *t2 = equalize_date_string (s2);
	int ret = strcmp (t1, t2);
	free (t1);
	free (t2);
	return ret;
}

/* does NOT care of month and year! */

int datestring_to_timestamp (const char *s) {
	if (safe_strlen (s) < 19) {
		return 0;
	}
	int ret = (s[8] - '0') * 10;
	ret += s[9] - '0';
	ret *= 24;
	ret += (s[11] - '0') * 10;
	ret += s[12] - '0';
	ret *= 60;
	ret += (s[14] - '0') * 10;
	ret += s[15] - '0';
	ret *= 60;
	ret += (s[17] - '0') * 10;
	ret += s[18] - '0';
	if (ret < 0) {
		return 0;
	}
	return ret;
}

